File: src/csharp/Intel/Iced/Intel/UsedRegister.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if INSTR_INFO
namespace Iced.Intel {
	/// <summary>
	/// A register used by an instruction
	/// </summary>
	public readonly struct UsedRegister {
		readonly Register register;
		readonly OpAccess access;

		/// <summary>
		/// Register
		/// </summary>
		public Register Register => register;

		/// <summary>
		/// Register access
		/// </summary>
		public OpAccess Access => access;

		/// <summary>
		/// Constructor
		/// </summary>
		/// <param name="register">Register</param>
		/// <param name="access">Register access</param>
		public UsedRegister(Register register, OpAccess access) {
			this.register = register;
			this.access = access;
		}

		/// <summary>
		/// ToString()
		/// </summary>
		/// <returns></returns>
		public override string ToString() => Register.ToString() + ":" + Access.ToString();
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/info/UsedRegister.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.info;

/**
 * A register used by an instruction
 */
public final class UsedRegister {
	private final byte register;
	private final byte access;

	/**
	 * Register (a {@link com.github.icedland.iced.x86.Register} enum variant)
	 */
	@SuppressWarnings("deprecation")
	public int getRegister() {
		return register & com.github.icedland.iced.x86.internal.Constants.REG_MASK;
	}

	/**
	 * Register access (an {@link com.github.icedland.iced.x86.info.OpAccess} enum variant)
	 */
	public int getAccess() {
		return access;
	}

	/**
	 * Returns a copy of this instance
	 */
	public UsedRegister copy() {
		return new UsedRegister(getRegister(), getAccess());
	}

	/**
	 * Constructor
	 *
	 * @param register Register (a {@link com.github.icedland.iced.x86.Register} enum variant)
	 * @param access   Register access (an {@link com.github.icedland.iced.x86.info.OpAccess} enum variant)
	 */
	@SuppressWarnings("deprecation")
	UsedRegister(int register, int access) {
		assert Integer.compareUnsigned(register, com.github.icedland.iced.x86.internal.Constants.REG_MASK) <= 0 : register;
		this.register = (byte)register;
		this.access = (byte)access;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + access;
		result = prime * result + register;
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		UsedRegister other = (UsedRegister)obj;
		return access == other.access && register == other.register;
	}
}
